types/index.d.ts

```diff
@@ -7,80 +7,188 @@
 import * as stream from 'stream';
 import { EventEmitter } from 'events'
 
-type Schema = string | object;  // TODO object should be further specified
+//"virtual" namespace (no JS, just types) for Avro Schema
+declare namespace schema
+{
+  export type AvroSchema = DefinedType | DefinedType[];
+  type DefinedType = PrimitiveType | ComplexType | LogicalType; 
+  type PrimitiveType = 'null' | 'boolean' | 'int' | 'long' | 'float' | 'double' | 'bytes' | 'string';
+  type ComplexType = NamedType | RecordType | EnumType | MapType | ArrayType | FixedType;
+  type LogicalType = ComplexType & LogicalTypeExtension;
+  
+  interface NamedType
+  {
+    type: PrimitiveType
+  }
+  
+  interface RecordType
+  {
+    type: "record";
+    name: string;
+    namespace? : string;
+    doc?: string;
+    aliases?: string[];
+    fields: {
+          name: string;
+          doc?:string;
+          type: AvroSchema;
```

Comment:
All nested `AvroSchema`s could be relaxed to `Schema` (`Type`s are also accepted).

Reply:
Schema lives in a namespace outside this one (one level up).  Would that not violate layering? I thought to make this namespace completely independent on how it is being used. It is just a typescript representation of the Avro spec. Or am I misunderstanding you?

Reply:
It's sometimes convenient to nest types inside of a schema declaration (a simpler alternative to passing them through `Type.forSchema`'s registry option). I'm not sure whether the namespacing is worth the extra restriction; it might also break existing code.

Reply:
Will change.